### DISCUSSION TITLE
Including Shadow in larger federated simulations

### DISCUSSION THREAD

**Discussion by hammer498** [shadow/shadow#2689] [Q&A] [answered]
We are investigating gluing shadow into a larger federated simulation

We have an existing federated simulation which is too large to fit on a single host.  Shadow looks like a promising tool to help onboard new code to simulation since it lets us control time in applications with no app changes. This control would only be useful to us however if we can control time and some messaging externally.  Our current theory for how this could work is
- Make a “gateway” app that runs in shadow alongside the app we want to add to our federation
	- The gateway wakes up periodically (in sim time) to disable interposition and check an external socket.  Our federation can then ask the gateway to block, or advance sim time to some specific point.
	- The gateway also proxies some requests (likely http) outside of shadow to the rest of the federation.
	- True federation-wide determinism isn’t a requirement but it is a great perk that apps inside shadow would get it

We are imagining this requires accessing real sockets outside of emulation but we don’t need any real threads.  At the moment we are getting more familiar with shadow and [experimenting with disabling interposition](https://github.com/shadow/shadow/discussions/2685) to see what does/doesn’t work. 

More context is available [in this video](https://www.youtube.com/watch?v=IlFdV424vyQ)

Would love to hear your thoughts or advice.  Do you think this is a plausible goal?

**Comment by robgjansen** [comment, marked as answer]
Hello! Thanks for your intriguing idea for using Shadow in innovative ways to accomplish broad experimentation goals!

Overall, although the devil is in the details, we think that it seems plausible that you could build a management program that runs in Shadow that periodically disables Shadow's interposition in order to connect or do some socket operations out to non-Shadow components, and then reenable Shadow's interposition and resume the Shadow simulation.

>  Our federation can then ask the gateway to block, or advance sim time to some specific point.

Shadow doesn't have an API to "advance sim time" for a whole host or simulation. It sounds like you might be able to accomplish what you want though by having your management program wait until a given simulation time (e.g. via a simulated call to `nanosleep`, or using something like `timerfd`) before sending proxied responses to the child simulators, or maybe even have an explicit protocol with the child simulators to tell them to wait until a given simulation time.

Also, we're not sure what happens if Shadow is not able to meet the real-time requirements of the federated parts of your simulation (e.g., if Shadow's load is too high that it is unable to advance simulation time fast enough), but we assume you have a good handle on those parts and have mechanisms to deal with that (or are running few-enough nodes in Shadow so as to not be worried about this).

**Reply by hammer498** [reply]
Thanks for the reply!
- Yes a simulated call to sleep or equivalent is what we were imagining
- You are absolutely right that having real time components in a federation makes life awkward for everyone.  Part of the appeal of shadow is it could help us convert some of those only-at-real-time components into simulated time components

If we run into any of those devils in the details we will keep reaching out.  Really appreciate how active and responsive this project is

**Reply by robgjansen** [reply]
Great! Thanks, and looking forward to seeing you succeed :)
